python/ray/dag/compiled_dag_node.py

```diff
@@ -867,6 +876,81 @@ def _get_node_id(self):
 
         self._dag_submitter.start()
         self._dag_output_fetcher.start()
+        self._max_concurrent_executions = self._compute_max_concurrent_executions()
+
+    @property
+    def max_concurrent_executions(self) -> int:
+        """
+        Get the maximum number of concurrent executions that the DAG
+        can perform without blocking.
+        """
+        return self._max_concurrent_executions
+
+    def _compute_max_concurrent_executions(self):
+        """
+        Compute the max possible concurrent executions of the DAG. This is the
+        number of inputs the DAG could start processing without blocking when
+        none of the outputs is consumed.
+
+        Returns:
+            The max possible concurrent executions of the DAG.
+
+        NOTE: The max possible concurrent executions is determined by the
+        "capacity" of the DAG, i.e., the number of intermediate results the
+        DAG can hold. The DAG holds these intermediate results in either
+        the intermediate channels or the actors, specifically:
+        - Channel capacity: either 1 or 0 depending on the channel type.
+        - Actor capacity: capacity is 1 because the actor can fetch and hold one
+            object from each input channel during execution. For an actor with
+            multiple methods, the capacity is still 1 because the current method
+            needs to finish (which flushes the held object) before the next method
+            can start.
+
+        This method performs a modified Dijkstra's algorithm to compute the
+        capacity of the DAG.
+        """
+        capa = {idx: float("inf") for idx in self.idx_to_task}
+        priority_queue = []
+
+        # Initial capacity is 1 as input channel has capacity 1.
+        capa[self.input_task_idx] = 1
```

Comment:
The comment says that the input channel's capacity is 1, but we also calculate the channel capacity again:

```python
chan_capa = self.idx_to_task[cur_idx].output_channel.capacity()
```

It is correct because we skip the output channel's (actor -> driver) capacity. The logic is not easy to follow. Maybe we can do some changes:

1. 
```python
capa[self.input_task_idx] = 0
heapq.heappush(priority_queue, (0, self.input_task_idx, set()))
```

2. Move `if cur_idx == self.output_task_idx:` right before `heappush`.

Reply:
> It is correct because we skip the output channel's (actor -> driver) capacity. 
not sure if I understand this.

My original logic had `capa[self.input_task_idx] = 0`, I changed it and set to 1 to be consistent with `downstream_capa = cur_capa + actor_capa + chan_capa`: it would be more natural to implement and read.

I also updated comments, hopefully now it is clearer to follow.